sunscreen/src/types/bfv/unsigned.rs

```diff
@@ -0,0 +1,358 @@
+
+        let sig_bits = self.val.bits_vartime();
+        seal_plaintext.resize(sig_bits);
+
+        for i in 0..sig_bits {
+            let bit_value = self.val.bit_vartime(i);
+            seal_plaintext.set_coefficient(i, bit_value);
+        }
+
+        Ok(Plaintext {
+            data_type: self.type_name_instance(),
+            inner: InnerPlaintext::Seal(vec![WithContext {
+                params: params.clone(),
+                data: seal_plaintext,
+            }]),
+        })
+    }
+}
+
+impl<const LIMBS: usize> TryFromPlaintext for Unsigned<LIMBS> {
+    fn try_from_plaintext(
+        plaintext: &Plaintext,
+        params: &Params,
+    ) -> std::result::Result<Self, sunscreen_runtime::Error> {
+        let val = match &plaintext.inner {
+            InnerPlaintext::Seal(p) => {
+                if p.len() != 1 {
+                    return Err(sunscreen_runtime::Error::IncorrectCiphertextCount);
+                }
+
+                let bits = usize::min(std::mem::size_of::<UInt<LIMBS>>() * 8, p[0].len());
+
+                let negative_cutoff = (params.plain_modulus + 1) / 2;
+
+                let mut val = UInt::ZERO;
+                for i in 0..bits {
+                    let coeff = p[0].get_coefficient(i);
+                    if coeff < negative_cutoff {
+                        val = wrapping_add(
+                            val,
+                            wrapping_mul(UInt::from_u8(0x1) << i, UInt::from_u64(coeff)),
+                        );
+                    } else {
+                        val = wrapping_sub(
+                            val,
+                            wrapping_mul(
+                                UInt::from_u8(0x1) << i,
+                                UInt::from_u64(params.plain_modulus - coeff),
+                            ),
+                        );
+                    }
+                }
+
+                Self { val }
+            }
+        };
+
+        Ok(val)
+    }
+}
+
+impl<const LIMBS: usize> From<UInt<LIMBS>> for Unsigned<LIMBS> {
+    fn from(val: UInt<LIMBS>) -> Self {
+        Self { val }
+    }
+}
+
+impl<const LIMBS: usize> From<u64> for Unsigned<LIMBS> {
+    fn from(n: u64) -> Self {
+        Self {
+            val: UInt::from_u64(n),
+        }
+    }
+}
+
+impl<const LIMBS: usize> From<Unsigned<LIMBS>> for UInt<LIMBS> {
+    fn from(unsigned: Unsigned<LIMBS>) -> Self {
+        unsigned.val
+    }
+}
+
+fn wrapping_add<const LIMBS: usize>(lhs: UInt<LIMBS>, rhs: UInt<LIMBS>) -> UInt<LIMBS> {
+    (Wrapping(lhs) + Wrapping(rhs)).0
+}
+fn wrapping_mul<const LIMBS: usize>(lhs: UInt<LIMBS>, rhs: UInt<LIMBS>) -> UInt<LIMBS> {
+    (Wrapping(lhs) * Wrapping(rhs)).0
+}
+fn wrapping_sub<const LIMBS: usize>(lhs: UInt<LIMBS>, rhs: UInt<LIMBS>) -> UInt<LIMBS> {
+    (Wrapping(lhs) - Wrapping(rhs)).0
+}
+
+macro_rules! impl_std_op {
+    ($($op:ident),+) => {
+        $(
+            paste! {
+                impl<const LIMBS: usize> $op for Unsigned<LIMBS> {
```

Comment:
I never did figure out how to wrote macro_rules! and had to use proc_macro, which entails a ton of shittiness.

A thing I've always wanted was a macro that would imlement each combination of borrow/owned for lhs and rhs in std ops. I think this would even compose well with such a macro!

Nothing to do here.